File: elpizo/models/geometry.py
from elpizo.models import record
from elpizo.protos import geometry_pb2
from elpizo.util import geometry


class Vector2(geometry.Vector2, record.ProtobufRecord):
  PROTOBUF_TYPE = geometry_pb2.Vector2
  FIELDS = [
      record.Field("x", record.Scalar),
      record.Field("y", record.Scalar),
  ]

  def __init__(self, x=None, y=None):
    # We need to make the x and y parameters optional for deserialization to
    # work.
    super().__init__(x, y)

  def __repr__(self):
    return "models.geometry." + super().__repr__()


class Rectangle(geometry.Rectangle, record.ProtobufRecord):
  PROTOBUF_TYPE = geometry_pb2.Rectangle
  FIELDS = [
      record.Field("left", record.Scalar),
      record.Field("top", record.Scalar),
      record.Field("width", record.Scalar),
      record.Field("height", record.Scalar),
  ]


  def __init__(self, left=None, top=None, width=None, height=None):
    # We need to make all the parameters optional for deserialization to work.
    super().__init__(left, top, width, height)

  def __repr__(self):
    return "models.geometry." + super().__repr__()

File: client/models/geometry.js
module geometry from "client/util/geometry";
module geometryProtos from "client/protos/geometry";

export class Vector2 extends geometry.Vector2 {
  toProtobuf() {
    return new geometryProtos.Vector2({x: this.x, y: this.y});
  }

  copy() {
    return new Vector2(this.x, this.y);
  }
}

Vector2.fromProtobuf = (proto) => {
  return new Vector2(proto.x, proto.y);
}

export class Vector3 extends geometry.Vector3 {
  toProtobuf() {
    return new geometryProtos.Vector3({x: this.x, y: this.y, z: this.z});
  }

  copy() {
    return new Vector3(this.x, this.y, this.z);
  }
}

Vector3.fromProtobuf = (proto) => {
  return new Vector3(proto.x, proto.y, proto.z);
}

export class Rectangle extends geometry.Rectangle {
  toProtobuf() {
    return new geometryProtos.Rectangle({left: this.left, top: this.top,
                                         width: this.width,
                                         height: this.height});
  }

  copy() {
    return new Rectangle(this.left, this.top, this.width, this.height);
  }
}

Rectangle.fromProtobuf = (proto) => {
  return new Rectangle(proto.left, proto.top, proto.width, proto.height);
}
